talk/tools/profiling.tex

```diff
@@ -4,39 +4,228 @@
+   stat            Run command and gather performance counter statistics
+   top             System profiling tool.
+   version         display the version of perf binary
+   probe           Define new dynamic tracepoints
+   trace           strace inspired tool
+ See 'perf help COMMAND' for more information on a specific command.
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Listing events with \mintinline{bash}{perf list}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # List main hardware events
+$ perf list hw
+
+List of pre-defined events (to be used in -e):
+
+  branch-instructions OR branches                    [Hardware event]
+  branch-misses                                      [Hardware event]
+  cache-misses                                       [Hardware event]
+  cache-references                                   [Hardware event]
+  cpu-cycles OR cycles                               [Hardware event]
+  instructions                                       [Hardware event]
+
+$ # List main software/cache events
+$ perf list sw
+$ perf list cache
+
+$ # List all pre-defined metrics
+$ perf list metric
+
+$ # List all currently known events:
+$ perf list
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Counting events with \mintinline{bash}{perf stat}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # Standard CPU counter statistics for the specified command:
+$ perf stat <command>
+
+$ # Detailed CPU counter statistics for the specified command:
+$ perf stat -d <command>
+$ perf stat -dd <command>
+
+$ # Top-down microarchitecture analysis for the entire system, for 10s:
+$ perf stat -a --topdown -- sleep 10
+
+$ # L1 cache hit rate reported every 1000 ms for the specified command:
+$ perf stat -e L1-dcache-loads,L1-dcache-load-misses -I 1000 <command>
+
+$ # Instruction per cycle and Instruction-level parallelism, for command:
+$ perf stat -M IPC,ILP -- <command>
+
+$ # Measure GFLOPs system-wide, until Ctrl-C is used to stop:
+$ perf stat -M GFLOPs
+
+$ # Measure cycles and instructions 10 times, report results with stddev:
+$ perf stat -e cycles,instructions -r 10 -- <command>
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitle{Recording profiling information with \mintinline{bash}{perf record}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # Sample on-CPU functions for the specified command, at 100 Hertz:
+$ perf record -F 100 -- <command>
+
+$ # Sample CPU stack traces (via frame pointers), at 100 Hertz, for 10s:
+$ perf record -F 100 -g -- sleep 10
+
+$ # Sample stack traces for PID using DWARF to unwind stacks, for 10s:
+$ perf record -p <PID> --call-graph=dwarf -- sleep 10
+
+$ # Precise on-CPU user stack traces (no skid) using PEBS (Intel CPUs):
```

Comment:
I explained on-CPU above. Basically, there is a margin of error to attribute samples to instructions, as a number of instructions are in flight in parallel on the CPU at any given time. This error is called the *skid* in the sampling (see more information [here](https://easyperf.net/blog/2018/08/29/Understanding-performance-events-skid)). PEBS stands for Precise Event Based Sampling (PEBS), and is a feature on Intel CPUs that allows sampling with low or no *skid*. The sort of equivalent thing on AMD CPUs is IBS, or Instruction-based Sampling.

Reply:
> I am asking because a future presenter of these slides might not know this. Is all the information relevant?

I hope that someone presenting `perf` to others will read the manual pages and understand these examples ahead of time. I tried to give a general overview of how to do several different things with each of the most important commands, so of course that what I added I think is relevant information for people trying to use `perf`. Maybe this is all too complicated for a C++ course and we should really just point people to the actual documentation or other material instead. I'm starting to think that that will be easier.

Reply:
> Maybe this is all too complicated for a C++ course

Do we need a tool section in the expert part ? That could be a solution

Reply:
Maybe a tools course, separate from a C++ course. VTune, perf, valgrind, can all be used for much more than just C++, so we can bundle this together with bash, coreutils, and some other command line tools that are used very often and make a new course.